dali/kernels/shape.h

```diff
@@ -0,0 +1,299 @@
+
+  TensorShape() = default;
+  TensorShape(const TensorShape &) = default;
+  TensorShape(TensorShape &&) = default;
+  TensorShape &operator=(const TensorShape &other) = default;
+  TensorShape &operator=(TensorShape &&other) = default;
+
+  // TODO(klecki) efficient size calculation for this vector?
+  template <int other_dim>
+  TensorShape(const TensorShape<other_dim> &other) : shape(other.shape.begin(), other.shape.end()) {
+    static_assert(other_dim != DynamicTensorShape, "This constructor should be not used");
+  }
+
+  template <int other_dim>
+  TensorShape(TensorShape<other_dim> &&other) : shape(other.shape.begin(), other.shape.end()) {}
+
+  template <int other_dim>
+  TensorShape &operator=(const TensorShape<other_dim> &other) {
+    shape = std::vector<int64_t>(other.shape.begin(), other.shape.end());
+    return *this;
+  }
+
+  template <int other_dim>
+  TensorShape &operator=(TensorShape<other_dim> &&other) {
+    shape = std::vector<int64_t>(other.shape.begin(), other.shape.end());
+    return *this;
+  }
+
+  // template <int other_dim>
+  // TensorShape(const TensorShape<other_dim> &other) {
+  //   *this = other;
+  // }
+
+  // template <int other_dim>
+  // TensorShape &operator=(const TensorShape<other_dim> &other) {
+  //   data.resize(other.size());
+  //   for (int i = 0; i < size(); i++) {
+  //     data[i] = other[i];
+  //   }
+  //   return *this;
+  // }
+
+  // TensorShape &operator=(TensorShape<-1> &&other) {
+  //   data = std::move(other.data);
+  //   return *this;
+  // }
+
+  // We allow only explicit conversion from dynamic to static dim
+  // template <int other_dim>
+  // explicit operator TensorShape<other_dim>() const;
+  template <int other_dim>
+  TensorShape<other_dim> to_static() const;
+
+  int64_t &operator[](int d) { return shape[d]; }
+  const int64_t &operator[](int d) const { return shape[d]; }
+
+  std::vector<int64_t> shape;
+  int size() const { return shape.size(); }
+};
+
+template <int dim_>
+struct TensorShape {
+  TensorShape(const std::array<int64_t, dim_> &s) : shape(s) {
+    std::cout << "const std::array<int64_t, dim_> &s" << std::endl;
+  }
+  TensorShape(std::array<int64_t, dim_> &&s) : shape(std::move(s)) {
+    std::cout << "std::array<int64_t, dim_> &&s" << std::endl;
+  }
+
+  // We just want zero-initialization
+  TensorShape() : shape{} {}
+  // We allow only explicit operations on TensorShape static dim
+  TensorShape(const TensorShape &) = default;
+  TensorShape(TensorShape &&) = default;
+  TensorShape &operator=(const TensorShape &other) = default;
+  TensorShape &operator=(TensorShape &&other) = default;
+
+  template <typename... Ts>
+  TensorShape(int64_t i0, Ts... s) : shape{i0, int64_t{s}...} {
+    static_assert(sizeof...(Ts) == dim_ - 1, "Number of shapes passed must match dim_");
+    std::cout << "TensorShape(int64_t i0, Ts... s)" << std::endl;
+    // TODO static_assert for Ts == int64_t
```

Comment:
Don't static-assert that - we want implicit conversions, e.g. `{ 1, 2, 3, 4 }` is an implicit conversion from int...

Reply:
I think the `int64_t{s}...` is safe enough.